### PR TITLE
Make apt repository configurable

### PR DESCRIPTION
**Pull request by krawthekrow**
Solution for #168.

This solution uses environment variables to propagate the chosen apt repository throughout the building process, and uses a lookup table to find the associated signing key. It requires the developer to have the repository signing key in his GPG keyring. Reprepro updates is used to implement the default repository; this also requires the developer to have the default repository's public key in his GPG keyring (gpg2 specifically, since that's what reprepro uses). ~It currently points to the main homeworld repository, but I'd work on making that configurable.~

### CODE REVIEW

File: building/build-debs/common/package-build-helpers.sh [hunk 1]
```diff
@@ -7,9 +7,23 @@ then
 	exit 1
 fi
 
+if [ -z "${APT_REPOSITORY:-}" ]
+then
+	echo "Need to specify apt repository:"
+	echo "\$ export APT_REPOSITORY=<username>/<branch>"
```
Comment (celskeggs) [inline building/build-debs/common/package-build-helpers.sh]: You can use single quotes for this, instead of escaping.

File: building/build-debs/homeworld-admin-tools/src/packages.py [hunk 2]
```diff
@@ -149,5 +148,11 @@ def verified_download_full(package_list: tuple) -> dict:
     """Download all the packages from the specified list from the main repository, including verifying them.
 
     Returns a mapping of {package_name: (short_filename, package_bytes), ...}"""
-    verified_info = download_and_verify_package_list(REPO_URL)
-    return {package_name: download_package(package_name, verified_info) for package_name in package_list}
+    apt_repository = resource.get_resource("APT_REPOSITORY").decode().rstrip()
+    apt_url = "http://web.mit.edu/hyades/apt/" + apt_repository
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/src/packages.py]: The base URL should still be a constant, instead of inline.

File: building/build-debs/homeworld-apt-setup/build-package.sh [hunk 3]
```diff
@@ -2,4 +2,8 @@
 set -eu
 source ../common/package-build-helpers.sh
 
+sed "s;\\\${APT_REPOSITORY};${APT_REPOSITORY};g" < homeworld.sources.in > homeworld.sources
```
Comment (celskeggs) [inline building/build-debs/homeworld-apt-setup/build-package.sh]: Might be worth doing this in a way that doesn't require that a specific character is not present in the value of ${APT_REPOSITORY} -- or at least will properly error out in that case.

File: building/upload-debs/conf/updates [hunk 4]
```diff
@@ -0,0 +1,4 @@
+Name: homeworld
+Components: main
+Method: http://web.mit.edu/hyades/debian/
+VerifyRelease: 9EE78588
```
Comment (celskeggs) [inline building/upload-debs/conf/updates]: We should never use an eight-character hash to verify anything, if we can help it. Can we use a 16-character hash here?

File: building/upload-debs/rebuild.sh [hunk 5]
```diff
@@ -1,7 +1,25 @@
...
+
+sed "s;\\\${APT_SIGNING_KEY};${APT_SIGNING_KEY};g" < conf/distributions.in > conf/distributions
+
 cd "$(dirname "$0")"
-reprepro -Vb . includedeb homeworld ../build-debs/binaries/homeworld-*.deb
-rsync -av --progress --delete-delay ./dists ./pool /mit/hyades/debian/
+gpg --import default-repo-signing-key.gpg
+reprepro -Vb . update
+reprepro -Vb . includedeb homeworld ../build-debs/binaries/${APT_REPOSITORY}/homeworld-*.deb
+rsync -av --progress --delete-delay ./dists ./pool /mit/hyades/apt/${APT_REPOSITORY}/
```
Comment (celskeggs) [inline building/upload-debs/rebuild.sh]: You should make sure that you properly quote any substitutions like this.

File: building/upload-debs/rebuild.sh [hunk 6]
```diff
@@ -1,7 +1,25 @@
...
+	exit 1
+fi
+
+sed "s;\\\${APT_SIGNING_KEY};${APT_SIGNING_KEY};g" < conf/distributions.in > conf/distributions
+
 cd "$(dirname "$0")"
-reprepro -Vb . includedeb homeworld ../build-debs/binaries/homeworld-*.deb
-rsync -av --progress --delete-delay ./dists ./pool /mit/hyades/debian/
+gpg --import default-repo-signing-key.gpg
+reprepro -Vb . update
```
Comment (celskeggs) [inline building/upload-debs/rebuild.sh]: Does `update` correctly handle the situation where we're trying to publish to the base repository?

File: building/upload-debs/rebuild.sh [hunk 7]
```diff
@@ -1,7 +1,25 @@
...
+	echo "\$ export APT_SIGNING_KEY=<key identifier>"
+	exit 1
+fi
+
+sed "s;\\\${APT_SIGNING_KEY};${APT_SIGNING_KEY};g" < conf/distributions.in > conf/distributions
+
 cd "$(dirname "$0")"
-reprepro -Vb . includedeb homeworld ../build-debs/binaries/homeworld-*.deb
-rsync -av --progress --delete-delay ./dists ./pool /mit/hyades/debian/
+gpg --import default-repo-signing-key.gpg
```
Comment (celskeggs) [inline building/upload-debs/rebuild.sh]: I'm wary about automatically changing the state of the user's GPG keyring. Can you comment on whether this is necessary?

File: building/build-debs/homeworld-admin-tools/inner-build.sh [hunk 8]
```diff
@@ -18,6 +18,10 @@ then
     echo >>src/resources/GIT_VERSION
 fi
 
+echo -n "$APT_REPOSITORY" >>src/resources/APT_REPOSITORY
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/inner-build.sh]: Is there a reason to have this file lack a trailing newline?

File: building/upload-debs/conf/updates [hunk 9]
```diff
@@ -0,0 +1,4 @@
+Name: homeworld
+Components: main
+Method: http://web.mit.edu/hyades/debian/
```
Comment (celskeggs) [inline building/upload-debs/conf/updates]: We probably want to change the upstream branch from `/hyades/debian/` to `/hyades/apt/root/master/` or something. Currently, the modified rebuild.sh script can't actually publish to the upstream branch...

File: building/build-debs/common/package-build-helpers.sh [hunk 10]
```diff
@@ -7,9 +7,23 @@ then
...
+then
+	echo "Need to specify apt signing key:"
+	echo "\$ export APT_SIGNING_KEY=<key identifier>"
+	exit 1
+fi
+
 DVERSION="$(head -n 1 debian/changelog | cut -d '(' -f 2 | cut -d ')' -f 1)"
 VERSION="$(echo "$DVERSION" | cut -d '-' -f 1)"
-BIN=../binaries
+BIN=../binaries/${APT_REPOSITORY}
```
Comment (celskeggs) [inline building/build-debs/common/package-build-helpers.sh]: This substitution should probably be quoted.

File: building/upload-debs/conf/updates (hunk 9 above)
Comment (krawthekrow) [reply building/upload-debs/conf/updates]: Configuring upstream is pretty much just changing this updates file; I don't think it's necessary to integrate it into our build process. In fact, if I understand reprepro correctly, you can define multiple upstream repositories here under different names, then it'd just be a matter of changing the name in the "Update:" field in conf/distributions to decide which one to use.

Could you move /hyades/debian/ to /hyades/apt/root/master/? (It's currently empty.) I don't have the bits for that.

File: building/upload-debs/rebuild.sh (hunk 7 above)
Comment (krawthekrow) [reply building/upload-debs/rebuild.sh]: I'm also uncomfortable with this part. However, reprepro doesn't seem to have a mechanism for using a key file directly. Looking at the man pages, though, it looks like I can specify a temporary GnuPG home directory for reprepro to use; I'll try doing that.

File: building/upload-debs/rebuild.sh (hunk 6 above)
Comment (krawthekrow) [reply building/upload-debs/rebuild.sh]: In theory, it should -- it only pulls updates if the version number is less than your currently built version, so you would have the updated debs in your repository when you rsync them over. It would be better to test this though -- could you do it just before merge? I don't have the bits for that

File: building/upload-debs/rebuild.sh (hunk 5 above)
Comment (krawthekrow) [reply building/upload-debs/rebuild.sh]: My mistake, will fix.

File: building/build-debs/homeworld-apt-setup/build-package.sh (hunk 3 above)
Comment (krawthekrow) [reply building/build-debs/homeworld-apt-setup/build-package.sh]: It would be nice if Bash had a cleaner way to do substitutions in files. Is this a security concern, or are we just handling the possibility of having special characters in future apt branches? I could use something from https://stackoverflow.com/a/2705678/1263980, but it might be easier (and better) to just limit the characters allowed in apt branch names, like to [A-Za-z0-9_-].

File: building/build-debs/homeworld-admin-tools/inner-build.sh (hunk 8 above)
Comment (krawthekrow) [reply building/build-debs/homeworld-admin-tools/inner-build.sh]: There isn't, but only because rstrip is done on the Python side. I'll remove it the flag here.

File: building/build-debs/common/package-build-helpers.sh (hunk 1 above)
Comment (krawthekrow) [reply building/build-debs/common/package-build-helpers.sh]: Thanks! Forgot you could do that.

File: building/build-debs/homeworld-apt-setup/build-package.sh (hunk 3 above)
Comment (celskeggs) [reply building/build-debs/homeworld-apt-setup/build-package.sh]: Just limiting the characters sounds good to me.

File: building/upload-debs/conf/updates (hunk 9 above)
Comment (celskeggs) [reply building/upload-debs/conf/updates]: I've moved the folder, and left a symlink from /hyades/debian.

Your thoughts on configurable upstreams seem reasonable. Good enough for me.

File: building/upload-debs/rebuild.sh (hunk 7 above)
Comment (celskeggs) [reply building/upload-debs/rebuild.sh]: Both of those work. Using a temporary GnuPG directory is good. Verifying that the key is in the keyring and telling the developer to import it if not is also good.

File: building/upload-debs/rebuild.sh (hunk 6 above)
Comment (celskeggs) [reply building/upload-debs/rebuild.sh]: Yeah, I can test this once we're closer to merge.

Have we confirmed that `update` verifies signatures properly?

File: building/build-debs/homeworld-admin-tools/build-package.sh [hunk 11]
```diff
@@ -11,5 +11,10 @@ git rev-parse HEAD | tr -d '\n' >resources/GIT_VERSION
 [[ -n "$(git status --porcelain)" ]] && echo -n "-dirty" >>resources/GIT_VERSION
 echo >>resources/GIT_VERSION
 
+echo "${HOMEWORLD_APT_BRANCH}" >>resources/APT_BRANCH
+
+HOMEWORLD_APT_SIGNING_KEY="$(cd ../.. && source get-apt-signing-key.sh)"
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/build-package.sh]: Why does this need to be sourced? Why do we need to change directories to use it?

Any reason not to make it a function instead of an entire script?

File: building/build-debs/homeworld-admin-tools/build-package.sh [hunk 12]
```diff
@@ -11,5 +11,10 @@ git rev-parse HEAD | tr -d '\n' >resources/GIT_VERSION
...
+gpg --export "${HOMEWORLD_APT_SIGNING_KEY}" >>resources/homeworld-archive-keyring.gpg
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/build-package.sh]: Maybe we want to clean up this file automatically, just like GIT_VERSION?

File: building/build-debs/homeworld-admin-tools/debian/rules [hunk 13]
```diff
@@ -17,7 +17,7 @@ include /usr/share/dpkg/default.mk
 #export DEB_LDFLAGS_MAINT_APPEND = -Wl,--as-needed
 
 override_dh_auto_build:
-	VERSION="${DEB_VERSION_UPSTREAM}" ./inner-build.sh
+	VERSION="${DEB_VERSION_UPSTREAM}" HOMEWORLD_APT_BRANCH="${HOMEWORLD_APT_BRANCH}" ./inner-build.sh
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/debian/rules]: Huh. Why is this necessary?

File: building/build-debs/homeworld-admin-tools/src/packages.py [hunk 14]
```diff
@@ -149,5 +150,11 @@ def verified_download_full(package_list: tuple) -> dict:
...
-    verified_info = download_and_verify_package_list(REPO_URL)
-    return {package_name: download_package(package_name, verified_info) for package_name in package_list}
+    apt_branch = resource.get_resource("APT_BRANCH").decode().rstrip()
+    apt_url = APT_REPO_BASE + apt_branch
+    try:
+        verified_info = download_and_verify_package_list(apt_url)
+        return {package_name: download_package(package_name, verified_info) for package_name in package_list}
+    except urllib.error.HTTPError:
+        command.fail("unable to access apt repository",
+            "do you have an apt repository at %s?" % apt_url)
```
Comment (celskeggs) [inline building/build-debs/homeworld-admin-tools/src/packages.py]: Might be worth using consistent terminology: apt branch, not apt repository.

File: building/get-apt-signing-key.sh [hunk 15]
```diff
@@ -0,0 +1,21 @@
+while read branch key; do
+	if [ "${branch}" == "${HOMEWORLD_APT_BRANCH}" ] || [ "${branch}" == "$(echo ${HOMEWORLD_APT_BRANCH} | cut -d '/' -f 1)" ]; then
+		HOMEWORLD_APT_SIGNING_KEY="${key}"
+		break
+	fi
+done < signing-keys
```
Comment (celskeggs) [inline building/get-apt-signing-key.sh]: For the sake of readability, I think people usually wrap constructs in parentheses if they're going to have things redirected into them afterwards?

i.e.:

    (for x in y; do
        <code>
    done) > output-file

File: building/get-apt-signing-key.sh [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+
+if [ -z "${HOMEWORLD_APT_SIGNING_KEY:-}" ]
+then
+	echo 'apt branch not found in signing-keys'
```
Comment (celskeggs) [inline building/get-apt-signing-key.sh]: Since the result is normally echoed, it might be worth making sure that errors are sent to stderr? (`1>&2`)

File: building/validate-apt-branch.sh [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+	exit 1
+fi
+
+apt_branch_regex='^[0-9a-zA-Z_-]+\/[0-9a-zA-Z_-]+$'
+if ! [[ "${HOMEWORLD_APT_BRANCH}" =~ $apt_branch_regex ]]
+then
+	echo 'Apt branch invalid. Should be of the form <username>/<branch>.'
+	echo 'Allowed characters: [0-9a-zA-Z_-]'
+	exit 1
+fi
```
Comment (celskeggs) [inline building/validate-apt-branch.sh]: I might consider turning get-apt-signing-key.sh into a function in this file.

File: docs/build.md [hunk 18]
```diff
@@ -17,7 +17,26 @@ That will let you build everything as described here.  If you also want to be ab
  * python3-yaml
  * zip
 
+# Create a personal apt repository
+
+Set the apt branch environment variable (you will have to do this every session):
+
+    $ export HOMEWORLD_APT_BRANCH=<username>/<branch>
+
+Do `gpg --list-keys --keyid-format long` to find the key you would like to sign your repository with. Add an entry to `building/signing-keys`:
```
Comment (celskeggs) [inline docs/build.md]: Perhaps suggest generating a key, and specifically discourage from using a regular long-lived PGP key?

File: docs/build.md [hunk 19]
```diff
@@ -17,7 +17,26 @@ That will let you build everything as described here.  If you also want to be ab
...
+
+    $ echo "${HOMEWORLD_APT_BRANCH} <key-id>" >> building/signing-keys
+
+Import the default apt repository signing key:
+
+    $ cd building/upload-debs
+    $ gpg --import default-repo-signing-key.gpg
+    $ cd ../..
```
Comment (celskeggs) [inline docs/build.md]: You can just have them import a relative path, instead of cding there.

File: building/validate-apt-branch.sh [hunk 20]
```diff
@@ -0,0 +1,14 @@
+if [ -z "${HOMEWORLD_APT_BRANCH:-}" ]
```
Comment (celskeggs) [inline building/validate-apt-branch.sh]: It might be worth sticking these two script in a subdirectory of `/building/`, as directly under `/building/` is a strange place to stick scripts.

File: building/build-debs/homeworld-admin-tools/debian/rules (hunk 13 above)
Comment (krawthekrow) [reply building/build-debs/homeworld-admin-tools/debian/rules]: Oops, forgot to remove that.

File: building/setup-apt-branch/setup-apt-branch.sh [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+    done < `dirname $BASH_SOURCE`/signing-keys
+
+    if [ -z "${HOMEWORLD_APT_SIGNING_KEY:-}" ]
+    then
+        echo 'error: apt branch not found in signing-keys' >&2
+        exit 1
+    fi
+
+    apt_signing_key_regex='^[0-9a-zA-Z]+$'
+    if ! [[ "${HOMEWORLD_APT_SIGNING_KEY}" =~ $apt_signing_key_regex ]]
```
Comment (celskeggs) [inline building/setup-apt-branch/setup-apt-branch.sh]: Is there a reason why you're defining a variable, just to use it exactly once?

File: building/setup-apt-branch/setup-apt-branch.sh [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    then
+        echo 'error: apt signing key invalid' >&2
+        exit 1
```
Comment (celskeggs) [inline building/setup-apt-branch/setup-apt-branch.sh]: Have you confirmed that this error checking works?

File: docs/build.md [hunk 23]
```diff
@@ -42,6 +63,8 @@ To upload packages, ONLY IF YOU ARE SUPPOSED TO RELEASE YOUR CHANGES:
 
     $ cd building/upload-debs
     $ aklog sipb    # authenticate to the SIPB AFS cell, if necessary
+        # Note that you will need Kerberos tickets
+        # (generate them with kinit) to access AFS.
```
Comment (celskeggs) [inline docs/build.md]: Any reason that one of these is after the 'aklog' and one of these is before?

File: building/setup-apt-branch/setup-apt-branch.sh (hunk 22 above)
Comment (krawthekrow) [reply building/setup-apt-branch/setup-apt-branch.sh]: Yes, I have tested this manually with different inputs.

File: docs/build.md (hunk 23 above)
Comment (krawthekrow) [reply docs/build.md]: No, I'll fix that.

File: building/setup-apt-branch/setup-apt-branch.sh (hunk 21 above)
Comment (krawthekrow) [reply building/setup-apt-branch/setup-apt-branch.sh]: No. I had problems getting the inline regex to work earlier, but it works now that other bugs have been fixed. I'll change both regexes to inline.

**Comment by celskeggs** [comment]
Needs rebase.
